---
title: "FMT Lipids Paper Analyses Script"
author: "Abigail Gancz"
date: "2025-10-16"
description: "This code is used to produce the figures and analyses ofr 'Successful Fecal Microbiota Transplants in Post-antibiotic Treated Recurrent Clostridioides difficile Patients Induce Acylcarnitine and Sphingolipid Lipidomic Shifts', specifically the independent lipid interrogations, the lipid/BA interrogations, and the lipid/microbiome interrogations."
output: html_document
---

# Packages & Knit Options
```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)

using<-function(...) {
  libs<-unlist(list(...))
  req<-unlist(lapply(libs,require,character.only=TRUE))
  need<-libs[req==FALSE]
  if(length(need)>0){ 
    install.packages(need)
    lapply(need,require,character.only=TRUE)}}

#call desired packages, and install if not available 
using("ALDEx2", "Maaslin2", "RColorBrewer", "RcppEigen", "RcppParallel", "Rtsne", "ape", "devtools", "fantaxtic", "forcats", 
      "ggforce", "ggplot2", "ggpubr", "ggtext", "grid", "gridExtra", "lattice", "microViz", "phyloseq", "phylosmith", 
      "randomForest", "reshape2", "rstatix", "scales", "tidyverse", "units", "vegan", "readxl", "utils", "pheatmap", 
      "factoextra", "ggfortify", "ggrepel", "cluster", "lmerTest", "rmcorr", "xlsx", "writexl", "janitor")

#import biobakery phyloseq functions
source("https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/phyloseq_biobakery_functions.R")
#BiocManager::install("phyloseq")

#set working directory 
setwd("C:/Users/aganc/Desktop/Professional/Projects/FMT_Lipid_MCBA_Microbiome_Paper/Analysis")

#sessionInfo()

```


# Data and metadata import 
```{r data}

#import metadata
metadata <- readRDS("../Data/metadata/metadata_sorted.rds")

#import lipids
  #import annotation keys
  lipid_group_key <- read.csv("../Data/lipids/lipids_to_class_key.csv")
  #import baker lab sample ids 
  baker_labels <- read.csv("../Data/lipids/2024_09_lipids/lipids_sampleID.csv")
  #import lipids
  #these values are TIC (total ion count) normalized, which adjusts the intensities of a mass spec data so that the total sum of all intensities equals to one 
  #these values are weight normalized, meaning that they are corrected for the amount of sample injected 
  lipids_raw <- read.xlsx("../Data/lipids/20240830_bothModes_FMT_CL_M1_WeightandTICNorm_updated_2.xlsx",sheetIndex = 1, check.names =F) %>%
    slice(-1) %>%
    dplyr::rename("Lipid"="Sample") %>%
    rename_at(baker_labels$Rank, ~ baker_labels$CLIENT.IDENTIFIER) %>%
    dplyr::select(Lipid,metadata$sample_ID)
  
#import taxonomy 
  #import biobakery phyloseq functions
  merged_abundance_table <- read.delim("../Data/metaphlan/merged_abundance_table.txt", comment.char="#")
  #import merged abundance table as ps object (replaced _kneaddata_metaphlan_bugs_list out of df)
  ps <- metaphlan_2phyloseq("../Data/metaphlan/merged_abundance_table.txt")
  otu_table <- otu_table(ps)
  sample_names(otu_table)
  tax_table <- tax_table(ps)
  #format metadata
  rownames(metadata) <- metadata$sample_ID
  ps_metadata <- sample_data(metadata)
  rownames(ps_metadata) <- metadata$sample_ID
  sample_names(ps_metadata)
  #merge ps with metadata
  ps <- phyloseq(otu_table, tax_table, ps_metadata)
  
  #metadata for taxonomy
  tax_metadata <- merged_abundance_table %>%
    select(clade_name) %>%
    separate(clade_name,
             into = c("kingdom", "phylum", "class", "order", "family", "genus", "species"),
             sep = "[;|]", 
             fill = "right",
             remove = FALSE) %>%
    mutate(across(c(kingdom, phylum, class, order, family, genus, species),
                  ~str_remove(., "^[a-z]__")))

#import pathways
  #merged, stratified, cpm-norm pathways
  pathway_raw<-read_tsv("../Data/humann/pathabundance_merged_cpm.tsv") %>%
    rename_with(~str_remove(.,"_kneaddata_Abundance-CPM")) %>%
    rename("pathway"='# Pathway') %>%
    select(pathway,levels(metadata$sample_ID)) %>%
    mutate_at(2:46,~as.numeric(.)) %>%
    as.data.frame()

#import pathways category key
Pathways_Category_Key <- read_excel("C:/Users/aganc/Desktop/Professional/Projects/FMT_Lipid_MCBA_Microbiome_Paper/Data/humann/Pathways_Category_Key.xlsx")


```

#Lipid Pre-Processing (filteration and normalization)
```{r data}

#adjust input data frame and clean lipid names
lipids_raw_cleaned <- lipids_raw %>%
  as.data.frame() %>%
  dplyr::rename("feature" = 1) %>%
  column_to_rownames("feature") %>%
  mutate(across(everything(), as.numeric))

#create a log2 transformed, 1e6 scaled, minimum value imputated dataset          
lipids_log2_1e6scaled_minimputated <- lipids_raw_cleaned %>%
  as.data.frame() %>%  # Ensure row names are retained
  tibble::rownames_to_column(var = "Lipid") %>%  # Store row names in a column
  mutate(across(where(is.numeric), ~ . * 1e6)) %>%  # Multiply by 1,000,000
  filter(rowSums(select(., -Lipid)) != 0) %>%  # Remove all-zero rows while ignoring lipid column
  rowwise() %>%  
  mutate(across(where(is.numeric), ~ if_else(. <= 1, min(c_across(where(is.numeric))[c_across(where(is.numeric)) > 1], na.rm = TRUE), .))) %>%
  ungroup() %>%
  mutate(across(where(is.numeric), log2)) %>% #perform log2 transformation
  column_to_rownames(var = "Lipid")  # Restore row names correctly

```

#Taxonomy Pre-Processing 
```{r data}

#subset by taxonomic rank 
  #Subset species
  taxa_relab_species <- merged_abundance_table %>%
      filter(str_detect(clade_name,"s__")&!str_detect(clade_name,"t__")) %>%
      mutate(clade_name=str_extract(clade_name, "s__.*")) %>%
      mutate(clade_name=str_remove(clade_name,"s__")) %>% 
      remove_rownames %>% 
      column_to_rownames(var="clade_name") %>% 
      as.data.frame()
  #subset genus
  taxa_relab_genus <- merged_abundance_table %>%
      filter(str_detect(clade_name,"g__")&!str_detect(clade_name,"s__")) %>%
      mutate(clade_name=str_extract(clade_name, "g__.*")) %>%
      mutate(clade_name=str_remove(clade_name,"g__") )%>% 
      remove_rownames %>% 
      column_to_rownames(var="clade_name") %>% 
      as.data.frame()
  #subset family
  taxa_relab_family <- merged_abundance_table %>%
      filter(str_detect(clade_name,"f__")&!str_detect(clade_name,"g__")) %>%
      mutate(clade_name=str_extract(clade_name, "f__.*")) %>%
      mutate(clade_name=str_remove(clade_name,"f__")) %>% 
      remove_rownames %>% 
      column_to_rownames(var="clade_name") %>% 
      as.data.frame()
  #subset order
  taxa_relab_order <- merged_abundance_table %>%
      filter(str_detect(clade_name,"o__")&!str_detect(clade_name,"f__")) %>%
      mutate(clade_name=str_extract(clade_name, "o__.*")) %>%
      mutate(clade_name=str_remove(clade_name,"o__")) %>% 
      remove_rownames %>% 
      column_to_rownames(var="clade_name") %>% 
      as.data.frame()
  #subset class
  taxa_relab_class <- merged_abundance_table %>%
      filter(str_detect(clade_name,"c__")&!str_detect(clade_name,"o__")) %>%
      mutate(clade_name=str_extract(clade_name, "c__.*")) %>%
      mutate(clade_name=str_remove(clade_name,"c__")) %>% 
      remove_rownames %>% 
      column_to_rownames(var="clade_name") %>% 
      as.data.frame()
  #subset phylum
  taxa_relab_phylum <- merged_abundance_table %>%
      filter(str_detect(clade_name,"p__")&!str_detect(clade_name,"c__")) %>%
      mutate(clade_name=str_extract(clade_name, "p__.*")) %>%
      mutate(clade_name=str_remove(clade_name,"p__")) %>% 
      remove_rownames %>% 
      column_to_rownames(var="clade_name") %>% 
      as.data.frame()

#microbe-specific subsets 
  #subset taxa genus to just bacteroidaceae
  taxa_relab_bacteroidaceae_species <- merged_abundance_table %>%
    filter(
      str_detect(clade_name, "f__Bacteroidaceae"),
      str_detect(clade_name, "s__"),
      !str_detect(clade_name, "t__")) %>%
    mutate(species_name = str_extract(clade_name, "s__[^;]*")) %>%
    mutate(species_name = str_remove(species_name, "s__")) %>%
    remove_rownames() %>%
    column_to_rownames(var = "species_name") %>%
    as.data.frame() %>%
    select(-1)
  #create annotation df for bacteroidaceae species
  bacteroidaceae_genus_species_lookup <- merged_abundance_table %>%
    filter(
      str_detect(clade_name, "f__Bacteroidaceae"),
      str_detect(clade_name, "s__"),
      !str_detect(clade_name, "t__")) %>%
    mutate(
      Genus = str_extract(clade_name, "g__[^|]*") %>% str_remove("g__"),
      Species = str_extract(clade_name, "s__[^|]*") %>% str_remove("s__")) %>%
    select(Genus, Species) %>%
    distinct()
  #subset taxa to just enterobacteriaceae
  taxa_relab_enterobacteriaceae_species <- merged_abundance_table %>%
    filter(
      str_detect(clade_name, "f__Enterobacteriaceae"),
      str_detect(clade_name, "s__"),
      !str_detect(clade_name, "t__")) %>%
    mutate(species_name = str_extract(clade_name, "s__[^;]*")) %>%
    mutate(species_name = str_remove(species_name, "s__")) %>%
    remove_rownames() %>%
    column_to_rownames(var = "species_name") %>%
    as.data.frame()  %>% 
    select(-1)
  #create annotation df for enterobacteriacea species
  entero_genus_species_lookup <- merged_abundance_table %>%
    filter(
      str_detect(clade_name, "f__Enterobacteriaceae"),
      str_detect(clade_name, "s__"),
      !str_detect(clade_name, "t__")) %>%
    mutate(
      Genus = str_extract(clade_name, "g__[^|]*") %>% str_remove("g__"),
      Species = str_extract(clade_name, "s__[^|]*") %>% str_remove("s__")) %>%
    select(Genus, Species) %>%
    distinct()
  #subset taxa to just lachnospiraceae
  taxa_relab_lachnospiraceae_species <- merged_abundance_table %>%
    filter(
      str_detect(clade_name, "f__Lachnospiraceae"),
      str_detect(clade_name, "s__"),
      !str_detect(clade_name, "t__")) %>%
    mutate(species_name = str_extract(clade_name, "s__[^;]*")) %>%
    mutate(species_name = str_remove(species_name, "s__")) %>%
    remove_rownames() %>%
    column_to_rownames(var = "species_name") %>%
    as.data.frame()  %>% 
    select(-1)
  #create annotation df for Lachnospiraceae species
  lachno_genus_species_lookup <- merged_abundance_table %>%
    filter(
      str_detect(clade_name, "f__Lachnospiraceae"),
      str_detect(clade_name, "s__"),
      !str_detect(clade_name, "t__")) %>%
    mutate(
      Genus = str_extract(clade_name, "g__[^|]*") %>% str_remove("g__"),
      Species = str_extract(clade_name, "s__[^|]*") %>% str_remove("s__"),
      Genus = if_else(str_starts(Genus, "GGB"), "Other", Genus)) %>%
    select(Genus, Species) %>%
    distinct()

```

#Pathway Pre-Processing 
```{r data}

#remove unmapped and unintegrated rows
pathway_clean <- pathway_raw[!grepl("UNMAPPED|UNINTEGRATED", pathway_raw$pathway), ] 
#create unstratified df
pathway_clean_unstrat <- pathway_clean[!grepl("\\|", pathway_clean$pathway), ]
#filter 
  #unstrat
    #abundance
    mean_abundance <- rowMeans(pathway_clean_unstrat[ , -1])
    pathway_clean_unstrat_filtered <- pathway_clean_unstrat[mean_abundance > 1, ]
    #prevalence
    prevalence <- rowSums(pathway_clean_unstrat_filtered[ , -1] > 0)
    keep <- prevalence >= 0.05 * (ncol(pathway_clean_unstrat_filtered) - 1)
    pathway_clean_unstrat_filtered <- pathway_clean_unstrat_filtered[keep, ]
  #strat
    #abundance
    mean_abundance <- rowMeans(pathway_clean[ , -1])
    pathway_clean_filtered <- pathway_clean[mean_abundance > 1, ]
    #prevalence
    prevalence <- rowSums(pathway_clean_filtered[ , -1] > 0)
    keep <- prevalence >= 0.05 * (ncol(pathway_clean_filtered) - 1)
    pathway_clean_filtered <- pathway_clean_filtered[keep, ]
    
```


#Fig 1B: Lipid PCA Biplot
```{r data}

#define the number of biplot vectors to show
num_vectors <- 0  

#transpose the data
lipids_transposed <- t(lipids_log2_1e6scaled_minimputated) 
lipids_transposed <- as.data.frame(lipids_transposed) 
lipids_transposed$sample_ID <- rownames(lipids_transposed)  

#perform pca
pca_result <- prcomp(lipids_transposed[, -ncol(lipids_transposed)], scale. = FALSE) 

#extract pca scores and join metadata
pca_scores <- as.data.frame(pca_result$x)  
pca_scores$sample_ID <- lipids_transposed$sample_ID 
pca_scores <- pca_scores %>% 
  left_join(metadata, by = "sample_ID")  

#biplot vectors
  #extract pca loadings (variable contributions)
  pca_loadings <- as.data.frame(pca_result$rotation) 
  pca_loadings$Lipid <- rownames(pca_loadings)  
  #calculate combined contribution (magnitude of loading vector)
  pca_loadings <- pca_loadings %>%
    mutate(contribution = sqrt(PC1^2 + PC2^2)) 
  #select the top 'num_vectors' variables based on contribution
  top_loadings <- pca_loadings %>%
    arrange(desc(contribution)) %>%  
    slice_head(n = num_vectors) 
  #scale the loadings for better visualization
  scaling_factor <- max(abs(pca_scores$PC1), abs(pca_scores$PC2)) / max(abs(top_loadings$PC1), abs(top_loadings$PC2))
  top_loadings <- top_loadings %>%
    mutate(PC1 = PC1 * scaling_factor, PC2 = PC2 * scaling_factor)

#plot pca with biplot vectors
ggplot(pca_scores, aes(x = PC1, y = PC2, color = time)) + 
  geom_point(size = 3) +  
  scale_color_manual(labels = c("Pre", "2 Week", "2 Month", "6 Month"), values = c("#b03900","#a5c0a6","#719673","#374f38")) +
  ggtitle("PCA of Log2-transformed, 1e6-Scaled, MinImput Lipid Data") +
  xlab(paste0("PC1 (", round(100 * summary(pca_result)$importance[2, 1], 1), "% variance)")) +
  ylab(paste0("PC2 (", round(100 * summary(pca_result)$importance[2, 2], 1), "% variance)")) +
  stat_ellipse(level = 0.95, size = 1) + 
  geom_segment(data = top_loadings, aes(x = 0, y = 0, xend = PC1, yend = PC2), arrow = arrow(length = unit(0.2, "cm")), color = "darkgrey", alpha = 0.7) +
  geom_text_repel(data = top_loadings, aes(x = PC1, y = PC2, label = Lipid), size = 3, color = "black", max.overlaps = 20) +
  theme(strip.background=element_blank(),
        strip.text = element_text(face = "bold", size = 5),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        plot.title = element_text(hjust = 0.5, size = 15, face = "bold"), 
        legend.position = "right") + theme_bw()
  
#export figure
  #select dir
  output_dir <- "../Figures"
  #generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_Lipid_PCA_NoBiplot.svg")  
  #save the plot
  ggsave(
    filename = file_name,
    plot = plot,
    width = 8,     
    height = 6,  
    dpi = 300,   
    limitsize = FALSE) 

```

#Statistical Test: Lipids pre/post and time (pre, 2 week, 2 month, 6 month) signficance testing 
```{r data}

#Use adonis significance test to determine whether there is a difference in these pre/post
  #Convert 'pathway' column to rownames (pathways as rows)
  adonis_matrix <- lipids_log2_1e6scaled_minimputated
  #Transpose so samples are rows, pathways are columns
  adonis_matrix_t <- t(adonis_matrix)
  #Filter metadata to only samples in the matrix
  adonis_metadata <- metadata %>% filter(sample_ID %in% rownames(adonis_matrix_t))
  #Reorder metadata rows to match matrix rows
  adonis_metadata <- adonis_metadata %>% arrange(match(sample_ID, rownames(adonis_matrix_t)))
  #Confirm rows match exactly
  all(rownames(adonis_matrix_t) == adonis_metadata$sample_ID)  # Should be TRUE)
  #Run PERMANOVA
  adonis_result <- adonis2(adonis_matrix_t ~ prepost, data = adonis_metadata, method = "bray", strata = adonis_metadata$Recipient)
  print(adonis_result)

#Use adonis significance test to determine whether there is a difference in these over time
  #Convert 'pathway' column to rownames (pathways as rows)
  adonis_matrix <- lipids_log2_1e6scaled_minimputated
  #Transpose so samples are rows, pathways are columns
  adonis_matrix_t <- t(adonis_matrix)
  #Filter metadata to only samples in the matrix
  adonis_metadata <- metadata %>% filter(sample_ID %in% rownames(adonis_matrix_t))
  #Reorder metadata rows to match matrix rows
  adonis_metadata <- adonis_metadata %>% arrange(match(sample_ID, rownames(adonis_matrix_t)))
  #Confirm rows match exactly
  all(rownames(adonis_matrix_t) == adonis_metadata$sample_ID)  # Should be TRUE)
  #Run PERMANOVA
  adonis_result <- adonis2(adonis_matrix_t ~ time, data = adonis_metadata, method = "bray", strata = adonis_metadata$Recipient)
  print(adonis_result)

``` 

#Statistical Test: Mixed Effects Modeling to determine lipids that are significant against time 
```{r data}

#melt and filter data
  #filter the metadata
  filtered_metadata <- metadata %>%
    filter(time %in% c("Pre", "2 week", "2 month", "6 month"))
  #keep only the columns in the metabolite matrix that match filtered metadata
  data_long <- lipids_log2_1e6scaled_minimputated %>%
    `[`(, colnames(.) %in% filtered_metadata$sample_ID) %>%
    rownames_to_column("Lipid") %>%
    pivot_longer(cols = -Lipid, names_to = "sample_ID", values_to = "value") %>%
    left_join(filtered_metadata, by = "sample_ID")

#get the unique list of lipids
unique_lipids <- unique(data_long$Lipid)

#initialize a dataframe
results <- data.frame(matrix(ncol = 4, nrow =0))
colnames(results) <- c('lipid', 'p_val_2week', 'p_val_2month', 'p_val_6month')

#loop through each lipid
for(i in 1:length(unique_lipids)){
  # Subset the data for the current lipid
  data_sub <- subset(data_long, Lipid == unique_lipids[i])
  model <- lmerTest::lmer(value ~ time + (1|Recipient), data = data_sub)
  summary <- summary(model)
  p_val_2week <- summary$coefficients[['time2 week', 'Pr(>|t|)']]
  p_val_2month <- summary$coefficients[['time2 month', 'Pr(>|t|)']]
  p_val_6month <- summary$coefficients[['time6 month', 'Pr(>|t|)']]
  # Append results
  results[i, 'lipid'] <- unique_lipids[i]
  results[i, 'p_val_2week'] <- p_val_2week
  results[i, 'p_val_2month'] <- p_val_2month
  results[i, 'p_val_6month'] <- p_val_6month
}

# Correct the p-values
results$p_val_2week <- p.adjust(results$p_val_2week, method = 'BH')
results$p_val_2month <- p.adjust(results$p_val_2month, method = 'BH')
results$p_val_6month <- p.adjust(results$p_val_6month, method = 'BH')

#export df
#write.csv(results, file = "../Tables/MEM_Lipid_Results.csv", row.names = FALSE)

#subset significant lipids
sig_df <- results %>%
  filter(p_val_2week < 0.05 | p_val_2month < 0.05 | p_val_6month < 0.05)
sig_lipids <- sig_df$lipid

#create new sig df
lipids_log2_1e6scaled_minimputated_sig <- lipids_log2_1e6scaled_minimputated %>%
  filter(rownames(.) %in% sig_lipids)

```

#Fig 1C: Heatmap with Mixed Effect Modeling (time) significant lipids
```{r data}

#melt and filter data  
  #filter the metadata
  filtered_metadata <- metadata %>%
    filter(time %in% c("Pre", "2 week", "2 month", "6 month"))
  # Keep only the columns in the metabolite matrix that match filtered Sample_IDs
  data <- lipids_log2_1e6scaled_minimputated_sig %>%
    as.matrix() %>%
    `[`(, colnames(.) %in% filtered_metadata$sample_ID) %>%
    .[apply(., 1, function(x) sd(x, na.rm = TRUE) != 0), ] # Remove rows with zero variance
  
#perform hierarchical clustering
    hclust_res <- hclust(dist(data, method = 'euclidean'), method = "complete")
    #calculate silhouette score for different numbers of clusters
    silhouette_scores <- sapply(2:10, function(k) {
      clusters <- cutree(hclust_res, k)
      sil <- silhouette(clusters, dist(data))
      mean(sil[, 3])})
    #plot silhouette scores for different k 
    plot(2:10, silhouette_scores, type = "b", main = "Silhouette Scores", xlab = "Number of Clusters", ylab = "Average Silhouette Score")  

#extract row order (do after first pass)
  ordered_row_names <- rownames(data)[heatmap$tree_row$order]
  #perform hierarchical clustering manually to match pheatmap's clusters
  row_clusters <- cutree(as.hclust(heatmap$tree_row), k = 2)
  #separate row names by cluster
  clustered_rows <- split(ordered_row_names, row_clusters[heatmap$tree_row$order])
  #check the results
  print(clustered_rows)
  #define cluster lipids
  cluster_1_lipids <- clustered_rows[[1]]
  cluster_2_lipids <- clustered_rows[[2]]    
    
#define annotations and colors for initial plot
    #match the order of samples in the heatmap
    annotation <- data.frame(Week = metadata$time)
    annotation$Week <- factor(annotation$Week, levels = c("Pre", "2 week", "2 month", "6 month"))
    rownames(annotation) <- metadata$sample_ID
    annotation_order <- annotation %>%
      rownames_to_column("sample_ID") %>%
      left_join(filtered_metadata, by = "sample_ID")
    #row annotation    
      row_annotation <- data.frame(Lipid = rownames(data)) %>%
        left_join(lipid_group_key, by = "Lipid") %>%
        column_to_rownames(var = "Lipid") # Ensure the row names match your data
    #define colors for the lipid class annotations
      print(unique(row_annotation$Group))
    #define annotation colors
        annotation_colors <- list(Week = c(
                                "Pre" = "#b03900",   
                                "2 week" = "#a5c0a6",
                                "2 month" = "#719673",
                                "6 month" = "#374f38"), 
                                  Group = c(
                                  "AC" = "#D99E49",
                                  "FA" = "#AD6928",
                                  "FAHFA" = "#684119" ,
                                  "Cer"  = "#F4BCC1",
                                  "GM3" = "#F18F99",
                                  "HexCer" = "#E64151",
                                  "SM" = "#BA202D",
                                  "LPC" = "#899DA5",
                                  "LPE" = "#9FCADD",
                                  "PC" = "#0B4064",
                                  "PE" = "#025485",
                                  "PG" = "#3D8BCA",
                                  "MGDG" = "#187D3E",
                                  "SQDG"  = "#204C24"))
    ordered_samples <- annotation_order %>% 
      arrange(Week) %>% 
      pull(sample_ID)
    data <- data[, ordered_samples]  

#separate out list of lipids by clusters 
#assign original plot to variable
heatmap <- pheatmap(
  mat = data,                         
  annotation_col = annotation,       
  annotation_colors = annotation_colors,
  cluster_cols = FALSE,
  scale = "row",                      
  clustering_distance_rows = 'euclidean', 
  clustering_distance_cols = 'euclidean', 
  clustering_method = "complete",     
  cutree_rows = 2,  
  fontsize_row = 4,
  fontsize_col = 7,
  color = colorRampPalette(c("white", "#99CCFF", "#000066"))(50)) # Heatmap colors

#extract row order
ordered_row_names <- rownames(data)[heatmap$tree_row$order]
#perform hierarchical clustering manually to match pheatmap's clusters
row_clusters <- cutree(as.hclust(heatmap$tree_row), k = 2)
#separate row names by cluster
clustered_rows <- split(ordered_row_names, row_clusters[heatmap$tree_row$order])
#check the results
print(clustered_rows)
#define cluster lipids
cluster_1_lipids <- clustered_rows[[1]]
cluster_2_lipids <- clustered_rows[[2]]

#export clusters
#write.table(cluster_1_lipids , "../cluster_1_lipids.txt", sep = "\t", row.names = FALSE, quote = FALSE)
#write.table(cluster_2_lipids , "../cluster_2_lipids.txt", sep = "\t", row.names = FALSE, quote = FALSE)

#melt and filter data  
#filter the metadata 
filtered_metadata <- metadata %>%
  filter(time %in% c("Pre", "2 week", "2 month", "6 month"))
#keep only the columns in the metabolite matrix that match filtered data, and remove cluster 2
data <- lipids_log2_1e6scaled_minimputated_sig %>%
  as.matrix() %>%
  `[`(, colnames(.) %in% filtered_metadata$sample_ID) %>%
  .[apply(., 1, function(x) sd(x, na.rm = TRUE) != 0), ] 

#import cluster key
lipid_species_to_class_key <- read.csv("../Data/lipids/lipid_species_to_class_key.csv")

#plot using pheatmap
#row annotation    
row_annotation <- lipid_species_to_class_key %>%
  column_to_rownames(var = "lipid") %>%
  as.data.frame() # Ensure the row names match your data
#define colors for the lipid class annotations
print(unique(row_annotation$class))
row_annotation$class <- factor(row_annotation$class, 
                               levels = c("AC_cluster1",
                                          "FA_cluster1",
                                          "FAHFA_cluster1",
                                          "Cer_cluster1",
                                          "GM3_cluster1",
                                          "HexCer_cluster1",
                                          "SM_cluster1",
                                          "LPC_cluster1",
                                          "LPE_cluster1",
                                          "PC_cluster1",
                                          "PE_cluster1",
                                          "PG_cluster1",
                                          "MGDG_cluster1",
                                          "SQDG_cluster1",
                                          "AC_cluster2",
                                          "FA_cluster2",
                                          "Cer_cluster2",
                                          "PC_cluster2",
                                          "PC.P._cluster2",
                                          "PE_cluster2",
                                          "PE.P._cluster2",
                                          "PG_cluster2"))
#column annotation 
#match the order of samples in the heatmap
annotation <- data.frame(Week = metadata$time)
annotation$Week <- factor(annotation$Week, levels = c("Pre", "2 week", "2 month", "6 month"))
rownames(annotation) <- metadata$sample_ID
annotation_order <- annotation %>%
  rownames_to_column("sample_ID") %>%
  left_join(filtered_metadata, by = "sample_ID")
#define annotation colors
annotation_colors <- list(Week = c(
  "Pre" = "#b03900",   
  "2 week" = "#a5c0a6",
  "2 month" = "#719673",
  "6 month" = "#374f38"), 
  class = c(
    "AC_cluster1"     = "#d99e49",
  "FA_cluster1"     = "#ad6928",
  "FAHFA_cluster1"  = "#684119",
  "Cer_cluster1"    = "#e7c1dc",
  "GM3_cluster1"    = "#cca1ca",
  "HexCer_cluster1" = "#ac78b4",
  "SM_cluster1"     = "#8d51a0",
  "LPC_cluster1"    = "#899da5",
  "LPE_cluster1"    = "#9fcadd",
  "PC_cluster1"     = "#0b4064",
  "PE_cluster1"     = "#025485",
  "PG_cluster1"     = "#3d8bca",
  "MGDG_cluster1"   = "#eb9e78",
  "SQDG_cluster1"   = "#ec8347",
  "AC_cluster2"     = "#d99e49",
  "FA_cluster2"     = "#ad6928",
  "Cer_cluster2"    = "#e7c1dc",
  "PC_cluster2"     = "#0b4064",
  "PC.P._cluster2"  = "#0b4064", 
  "PE_cluster2"     = "#025485",
  "PE.P._cluster2"  = "#025485",  
  "PG_cluster2"     = "#3d8bca"))

#order col by Week 
ordered_samples <- annotation_order %>% 
  arrange(Week) %>% 
  pull(sample_ID)
data <- data[, ordered_samples] 
#reorder the rows of 'data' according to the factor levels in 'row_annotation$class'
  row_annotation <- row_annotation[rownames(data), , drop = FALSE]
  ordered_idx <- order(row_annotation$class)
  data  <- data[ordered_idx, ]
  row_annotation <- row_annotation[ordered_idx, , drop = FALSE]
  
#plot
plot <- pheatmap(
  mat = data,                        
  annotation_col = annotation,        
  annotation_row = row_annotation,
  annotation_colors = annotation_colors, 
  cluster_cols = TRUE,
  cluster_rows = FALSE,
  scale = "row",                     
  clustering_distance_cols = 'euclidean', 
  clustering_method = "complete",    
  fontsize_row = 4,
  fontsize_col = 7,
  color = colorRampPalette(c("white", "#99CCFF", "#000066"))(50)) 

#export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_MEM_Sig_Lipids_Heatmap_clustered_and_ordered.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = plot,  
    width = 8,   
    height = 6,  
    dpi = 300,    
    limitsize = FALSE)

```

#Fig 2A: Maaslin2 differential abundance analysis volcano plot (prepost)
```{r data}

#transform data  
data <- lipids_log2_1e6scaled_minimputated %>%
          mutate_all(~as.numeric(.)) %>%
          t() %>%
          as.data.frame() %>%
          rename_with(~sub("^X", "", .), everything())

#ensure data and metadata match
head(rownames(data))
head(rownames(metadata))
length(intersect(rownames(data), rownames(metadata)))
common_samples <- intersect(rownames(data), rownames(metadata))
data <- data[common_samples, ]
metadata <- metadata[common_samples, ]

#run maaslin for prepost
  #make a directory
  dir.create("../Analysis/maaslin2/lipids_log2_1e6scaled_minimputated", recursive = TRUE)  
  #run maaslin
  Maaslin2(
          input_data = data,
          output = "../Analysis/maaslin2/lipids_log2_1e6scaled_minimputated_prepost",
          input_metadata = metadata %>% column_to_rownames("sample_ID"),
          fixed_effects = c("prepost"), 
          random_effects = c("Recipient"),
          reference=c("prepost,Pre"),
          normalization="NONE",
          transform="NONE",
          analysis_method ="LM",
          correction="bonferroni",
          max_significance = 0.05)

#import results
maaslin_results <- read_tsv("../Analysis/maaslin2/lipids_log2_1e6scaled_minimputated_prepost/all_results.tsv")

# Add a column to indicate significance
maaslin_results <- maaslin_results %>%
  mutate(log_q_value = -log10(qval)) %>%
  mutate(significance = ifelse(qval < 0.05, "Significant", "Not Significant")) %>%
  mutate(association = case_when(
      (coef > 0 & qval < 0.05) ~ "Post-FMT",  
      (coef < 0 & qval < 0.05) ~ "Pre-FMT",  
      TRUE ~ "Neutral"))

# Create the volcano plot
plot <- ggplot(maaslin_results, aes(x = coef, y = log_q_value, color = association)) +
  geom_point(alpha = 1, size = 3)  +
  scale_color_manual(values = c("Post-FMT" = "#719673", "Pre-FMT" = "#b03900", "Neutral" = "#7f7f7f")) + 
  labs(x = "Effect Size", y = "-Log10(q-value)", title = "Volcano Plot of MaAsLin2 Results (Prepost: Post)") +
  theme(legend.position="bottom",
    strip.background=element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    plot.title = element_text(hjust = 0.5, size = 15, face = "bold")) + 
  theme(strip.text = element_text(face = "bold", size = 5)) +
    geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
    geom_text(data = subset(maaslin_results, qval < 0.05), aes(label = feature), nudge_x = 0.5, nudge_y = .4, size = 3, color = "black") + theme_bw()
  
#export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_Maaslin_DAA_Lipids_PrePost.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = plot, 
    width = 8,     
    height = 6,  
    dpi = 300,   
    limitsize = FALSE) 

```

#Fig 2B: Lipid Combined Boxplots (MEM signficiant AC) (prepost)
```{r data}

#filter data
data <- lipids_log2_1e6scaled_minimputated_sig %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  select(starts_with("AC")) %>%
  rownames_to_column("sample_ID")  %>%
  pivot_longer(cols = starts_with("AC"), names_to = "lipid", values_to = "value") %>%
  left_join(metadata, by = "sample_ID")

#horizontal plot
plot <- ggplot(data, aes(x = prepost, y = value, fill = prepost)) +
  geom_boxplot() +  # Boxplot
  theme_minimal() +
  labs(title = "Boxplot of AC Lipids Across Time", x = "Log2, 1e6 scaled, min-impuatated value", y = "Lipid") +
  facet_grid(. ~ lipid, scales = "free_x") +
  scale_fill_manual(values = c("#353535","#EAEAEA")) + 
  coord_cartesian(ylim = c(0, 18))  +
  scale_y_continuous(breaks = seq(0, 18, by = 3)) +
  theme(legend.position="bottom",
        strip.background=element_blank(),
        axis.text.x = element_blank(),
        #panel.grid.major = element_blank(),
        #panel.grid.minor = element_blank(),
        #panel.border = element_rect(),
        panel.spacing = unit(1, "lines"),
        plot.title = element_text(hjust = 0.5, size = 15, face = "bold")) +
    geom_jitter(shape = 21, color = "black", fill = "white", width = 0.1, size = 2)  # White dots with black outlines

plot

#export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_AC_Boxplots_Prepost.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = plot, 
    width = 12,    
    height = 6,
    dpi = 300,   
    limitsize = FALSE) 

```

#Fig 2C: Lipid Combined Boxplots (MEM signficiant Cer) (prepost)
```{r data}

#filter data
data <- lipids_log2_1e6scaled_minimputated_sig %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  select(starts_with("Cer")) %>%
  rownames_to_column("sample_ID")  %>%
  pivot_longer(cols = starts_with("Cer"), names_to = "lipid", values_to = "value") %>%
  left_join(metadata, by = "sample_ID")

#horizontal plot
plot <- ggplot(data, aes(x = prepost, y = value, fill = prepost)) +
  geom_boxplot() +  # Boxplot
  theme_minimal() +
  labs(title = "Boxplot of Cer Lipids Across Time", x = "Log2, 1e6 scaled, min-impuatated value", y = "Lipid") +
  facet_grid(. ~ lipid, scales = "free_x") +
  scale_fill_manual(values = c("#353535","#EAEAEA")) + 
  coord_cartesian(ylim = c(0, 18))  +
  scale_y_continuous(breaks = seq(0, 18, by = 3)) + 
  theme(legend.position="bottom",
        strip.background=element_blank(),
        axis.text.x = element_blank(),
        #panel.grid.major = element_blank(),
        #panel.grid.minor = element_blank(),
        #panel.border = element_rect(),
        panel.spacing = unit(1, "lines"),
        plot.title = element_text(hjust = 0.5, size = 15, face = "bold")) +
    geom_jitter(shape = 21, color = "black", fill = "white", width = 0.1, size = 2)  # White dots with black outlines

plot

#export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_Cer_Boxplots_prepost.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = plot,  # Use the combined plot
    width = 12,     # Width of the plot in inches
    height = 6,    # Height of the plot in inches
    dpi = 300,     # Resolution (dots per inch)
    limitsize = FALSE)  # Prevent clipping of larger legends or labels

```

#Supplemental Figure 1A: Lipid Combined Boxplots (MEM signficiant FA)
```{r data}

#filter data
data <- lipids_log2_1e6scaled_minimputated_sig %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  select(starts_with("FA")) %>%
  rownames_to_column("sample_ID")  %>%
  pivot_longer(cols = starts_with("FA"), names_to = "lipid", values_to = "value") %>%
  left_join(metadata, by = "sample_ID")

#horizontal plot
plot <- ggplot(data, aes(x = prepost, y = value, fill = prepost)) +
  geom_boxplot() +  # Boxplot
  theme_minimal() +
  labs(title = "Boxplot of FA Lipids Across Time", x = "Log2, 1e6 scaled, min-impuatated value", y = "Lipid") +
  facet_grid(. ~ lipid, scales = "free_x") +
  scale_fill_manual(values = c("#353535","#EAEAEA")) + 
  coord_cartesian(ylim = c(0, 18))  +
  scale_y_continuous(breaks = seq(0, 18, by = 3)) + 
  theme(legend.position="bottom",
        strip.background=element_blank(),
        axis.text.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        plot.title = element_text(hjust = 0.5, size = 15, face = "bold")) +
    geom_jitter(shape = 21, color = "black", fill = "white", width = 0.1, size = 2)  # White dots with black outlines

plot

#export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_FA_Boxplots_Over_Time.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = plot,  # Use the combined plot
    width = 12,     # Width of the plot in inches
    height = 6,    # Height of the plot in inches
    dpi = 300,     # Resolution (dots per inch)
    limitsize = FALSE)  # Prevent clipping of larger legends or labels

```

#Supplemental Figure 1B: Lipid Combined Boxplots (MEM signficiant FAHFA)
```{r data}

#filter data
data <- lipids_log2_1e6scaled_minimputated_sig %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  select(starts_with("FAHFA")) %>%
  rownames_to_column("sample_ID")  %>%
  pivot_longer(cols = starts_with("FA"), names_to = "lipid", values_to = "value") %>%
  left_join(metadata, by = "sample_ID")

#horizontal plot
plot <- ggplot(data, aes(x = prepost, y = value, fill = prepost)) +
  geom_boxplot() +  # Boxplot
  theme_minimal() +
  labs(title = "Boxplot of FAHFA Lipids Across Time", x = "Log2, 1e6 scaled, min-impuatated value", y = "Lipid") +
  facet_grid(. ~ lipid, scales = "free_x") +
   scale_fill_manual(values = c("#353535","#EAEAEA")) + 
  coord_cartesian(ylim = c(0, 18))  +
  scale_y_continuous(breaks = seq(0, 18, by = 3)) + 
  theme(legend.position="bottom",
        strip.background=element_blank(),
        axis.text.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        plot.title = element_text(hjust = 0.5, size = 15, face = "bold")) +
    geom_jitter(shape = 21, color = "black", fill = "white", width = 0.1, size = 2)  # White dots with black outlines

plot

#export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_FAHFA_Boxplots_Over_Time.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = plot,  # Use the combined plot
    width = 12,     # Width of the plot in inches
    height = 6,    # Height of the plot in inches
    dpi = 300,     # Resolution (dots per inch)
    limitsize = FALSE)  # Prevent clipping of larger legends or labels

```
#Supplemental Figure 2A: Lipid Combined Boxplots (MEM signficiant GM3)
```{r data}

#filter data
data <- lipids_log2_1e6scaled_minimputated_sig %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  select(starts_with("GM3")) %>%
  rownames_to_column("sample_ID")  %>%
  pivot_longer(cols = starts_with("GM3"), names_to = "lipid", values_to = "value") %>%
  left_join(metadata, by = "sample_ID")

#horizontal plot
plot <- ggplot(data, aes(x = prepost, y = value, fill = prepost)) +
  geom_boxplot() +  # Boxplot
  theme_minimal() +
  labs(title = "Boxplot of GM3 Lipids Across Time", x = "Log2, 1e6 scaled, min-impuatated value", y = "Lipid") +
  facet_grid(. ~ lipid, scales = "free_x") +
  scale_fill_manual(values = c("#353535","#EAEAEA")) + 
  coord_cartesian(ylim = c(0, 18))  +
  scale_y_continuous(breaks = seq(0, 18, by = 3)) + 
  theme(legend.position="bottom",
        strip.background=element_blank(),
        axis.text.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        plot.title = element_text(hjust = 0.5, size = 15, face = "bold")) +
    geom_jitter(shape = 21, color = "black", fill = "white", width = 0.1, size = 2)  # White dots with black outlines

plot

#export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_GM3_Boxplots_Over_Time.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = plot,  # Use the combined plot
    width = 12,     # Width of the plot in inches
    height = 6,    # Height of the plot in inches
    dpi = 300,     # Resolution (dots per inch)
    limitsize = FALSE)  # Prevent clipping of larger legends or labels

```
#Supplemental Figure 2B: Lipid Combined Boxplots (MEM signficiant HexCer)
```{r data}

#filter data
data <- lipids_log2_1e6scaled_minimputated_sig %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  select(starts_with("HexCer")) %>%
  rownames_to_column("sample_ID")  %>%
  pivot_longer(cols = starts_with("HexCer"), names_to = "lipid", values_to = "value") %>%
  left_join(metadata, by = "sample_ID")

#horizontal plot
plot <- ggplot(data, aes(x = prepost, y = value, fill = prepost)) +
  geom_boxplot() +  # Boxplot
  theme_minimal() +
  labs(title = "Boxplot of HexCer Lipids Across Time", x = "Log2, 1e6 scaled, min-impuatated value", y = "Lipid") +
  facet_grid(. ~ lipid, scales = "free_x") +
    scale_fill_manual(values = c("#353535","#EAEAEA")) + 
  coord_cartesian(ylim = c(0, 18))  +
  scale_y_continuous(breaks = seq(0, 18, by = 3)) + 
  theme(legend.position="bottom",
        strip.background=element_blank(),
        axis.text.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        plot.title = element_text(hjust = 0.5, size = 15, face = "bold")) +
    geom_jitter(shape = 21, color = "black", fill = "white", width = 0.1, size = 2)  # White dots with black outlines

plot

#export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_HexCer_Boxplots_Over_Time.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = plot,  # Use the combined plot
    width = 12,     # Width of the plot in inches
    height = 6,    # Height of the plot in inches
    dpi = 300,     # Resolution (dots per inch)
    limitsize = FALSE)  # Prevent clipping of larger legends or labels

```


#Supplemental Figure 2C:Lipid Combined Boxplots (MEM signficiant SM)
```{r data}

#filter data
data <- lipids_log2_1e6scaled_minimputated_sig %>%
  as.matrix() %>%
  t() %>%
  as.data.frame() %>%
  select(starts_with("SM")) %>%
  rownames_to_column("sample_ID")  %>%
  pivot_longer(cols = starts_with("SM"), names_to = "lipid", values_to = "value") %>%
  left_join(metadata, by = "sample_ID")

#horizontal plot
plot <- ggplot(data, aes(x = prepost, y = value, fill = prepost)) +
  geom_boxplot() +  # Boxplot
  theme_minimal() +
  labs(title = "Boxplot of SM Lipids Across Time", x = "Log2, 1e6 scaled, min-impuatated value", y = "Lipid") +
  facet_grid(. ~ lipid, scales = "free_x") +
  scale_fill_manual(values = c("#353535","#EAEAEA")) + 
  coord_cartesian(ylim = c(0, 18))  +
  scale_y_continuous(breaks = seq(0, 18, by = 3)) + 
  theme(legend.position="bottom",
        strip.background=element_blank(),
        axis.text.x = element_blank(),
        panel.spacing = unit(1, "lines"),
        plot.title = element_text(hjust = 0.5, size = 15, face = "bold")) +
    geom_jitter(shape = 21, color = "black", fill = "white", width = 0.1, size = 2)  # White dots with black outlines

plot

#export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_SM_Boxplots_Over_Time.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = plot,  # Use the combined plot
    width = 12,     # Width of the plot in inches
    height = 6,    # Height of the plot in inches
    dpi = 300,     # Resolution (dots per inch)
    limitsize = FALSE)  # Prevent clipping of larger legends or labels

```

#Statistical Test: Mixed Effects Modeling to determine sig family, genera, species
```{r data}

#Taxa (Families)
    #melt and filter data
      #filter the metadata 
      filtered_metadata <- metadata %>%
        filter(time %in% c("Pre", "2 week", "2 month", "6 month"))
      #keep only the columns in the metabolite matrix that match filtered metadata
      data_long <- taxa_relab_family %>%
        `[`(, colnames(.) %in% filtered_metadata$sample_ID) %>%
        rownames_to_column("family") %>%
        pivot_longer(cols = -family, names_to = "sample_ID", values_to = "value") %>%
        left_join(filtered_metadata, by = "sample_ID")
    #get the unique list 
    unique_microbes <- unique(data_long$family)
    #initialize a dataframe
    results <- data.frame(matrix(ncol = 4, nrow =0))
    colnames(results) <- c('family', 'p_val_2week', 'p_val_2month', 'p_val_6month')
    #loop through each metabolite
    for(i in 1:length(unique_microbes)){
      #subset the data 
      data_sub <- subset(data_long, family == unique_microbes[i])
      model <- lmerTest::lmer(value ~ time + (1|Recipient), data = data_sub)
      summary <- summary(model)
      p_val_2week <- summary$coefficients[['time2 week', 'Pr(>|t|)']]
      p_val_2month <- summary$coefficients[['time2 month', 'Pr(>|t|)']]
      p_val_6month <- summary$coefficients[['time6 month', 'Pr(>|t|)']]
      #append results
      results[i, 'family'] <- unique_microbes[i]
      results[i, 'p_val_2week'] <- p_val_2week
      results[i, 'p_val_2month'] <- p_val_2month
      results[i, 'p_val_6month'] <- p_val_6month}
    #correct the p-values
    results$p_val_2week <- p.adjust(results$p_val_2week, method = 'BH')
    results$p_val_2month <- p.adjust(results$p_val_2month, method = 'BH')
    results$p_val_6month <- p.adjust(results$p_val_6month, method = 'BH')
    #write.csv(sig_taxa_family, file = "results_output_table.csv", row.names = TRUE)
    #subset significant lipids
    sig_df <- results %>%
      filter(p_val_2week < 0.05 | p_val_2month < 0.05 | p_val_6month < 0.05)
    sig_microbes <- sig_df$family
    #create new sig df
    taxa_relab_family_sig <- taxa_relab_family %>% 
      filter(rownames(.) %in% sig_microbes)
    
#Taxa (Genus)
    #melt and filter data
      #filter the metadata 
      filtered_metadata <- metadata %>%
        filter(time %in% c("Pre", "2 week", "2 month", "6 month"))
      #keep only the columns in the metabolite matrix that match filtered metadata
      data_long <- taxa_relab_genus %>%
        `[`(, colnames(.) %in% filtered_metadata$sample_ID) %>%
        rownames_to_column("genus") %>%
        pivot_longer(cols = -genus, names_to = "sample_ID", values_to = "value") %>%
        left_join(filtered_metadata, by = "sample_ID")
    #get the unique list 
    unique_microbes <- unique(data_long$genus)
    #initialize a dataframe
    results <- data.frame(matrix(ncol = 4, nrow =0))
    colnames(results) <- c('genus', 'p_val_2week', 'p_val_2month', 'p_val_6month')
    #loop through each metabolite
    for(i in 1:length(unique_microbes)){
      # Subset the data 
      data_sub <- subset(data_long, genus == unique_microbes[i])
      model <- lmerTest::lmer(value ~ time + (1|Recipient), data = data_sub)
      summary <- summary(model)
      p_val_2week <- summary$coefficients[['time2 week', 'Pr(>|t|)']]
      p_val_2month <- summary$coefficients[['time2 month', 'Pr(>|t|)']]
      p_val_6month <- summary$coefficients[['time6 month', 'Pr(>|t|)']]
      #append results
      results[i, 'genus'] <- unique_microbes[i]
      results[i, 'p_val_2week'] <- p_val_2week
      results[i, 'p_val_2month'] <- p_val_2month
      results[i, 'p_val_6month'] <- p_val_6month}
    #correct the p-values
    results$p_val_2week <- p.adjust(results$p_val_2week, method = 'BH')
    results$p_val_2month <- p.adjust(results$p_val_2month, method = 'BH')
    results$p_val_6month <- p.adjust(results$p_val_6month, method = 'BH')
    #write.csv(sig_taxa_family, file = "results_output_table.csv", row.names = TRUE)
    #subset significant lipids
    sig_df <- results %>%
      filter(p_val_2week < 0.05 | p_val_2month < 0.05 | p_val_6month < 0.05)
    sig_microbes <- sig_df$genus
    #create new sig df
    taxa_relab_genus_sig <- taxa_relab_genus %>% 
      filter(rownames(.) %in% sig_microbes)
    
#Taxa (Species)
    #melt and filter data
      #filter the metadata 
      filtered_metadata <- metadata %>%
        filter(time %in% c("Pre", "2 week", "2 month", "6 month"))
      #keep only the columns in the metabolite matrix that match filtered metadata
      data_long <- taxa_relab_species %>%
        `[`(, colnames(.) %in% filtered_metadata$sample_ID) %>%
        rownames_to_column("species") %>%
        pivot_longer(cols = -species, names_to = "sample_ID", values_to = "value") %>%
        left_join(filtered_metadata, by = "sample_ID")
    #get the unique list 
    unique_microbes <- unique(data_long$species)
    #initialize a dataframe
    results <- data.frame(matrix(ncol = 4, nrow =0))
    colnames(results) <- c('species', 'p_val_2week', 'p_val_2month', 'p_val_6month')
    #loop through each Metabolite
    for(i in 1:length(unique_microbes)){
      #subset the data for the current lipid
      data_sub <- subset(data_long, species == unique_microbes[i])
      model <- lmerTest::lmer(value ~ time + (1|Recipient), data = data_sub)
      summary <- summary(model)
      p_val_2week <- summary$coefficients[['time2 week', 'Pr(>|t|)']]
      p_val_2month <- summary$coefficients[['time2 month', 'Pr(>|t|)']]
      p_val_6month <- summary$coefficients[['time6 month', 'Pr(>|t|)']]
      #append results
      results[i, 'species'] <- unique_microbes[i]
      results[i, 'p_val_2week'] <- p_val_2week
      results[i, 'p_val_2month'] <- p_val_2month
      results[i, 'p_val_6month'] <- p_val_6month}
    #correct the p-values
    results$p_val_2week <- p.adjust(results$p_val_2week, method = 'BH')
    results$p_val_2month <- p.adjust(results$p_val_2month, method = 'BH')
    results$p_val_6month <- p.adjust(results$p_val_6month, method = 'BH')
    #write.csv(sig_taxa_family, file = "results_output_table.csv", row.names = TRUE)
    #subset significant lipids
    sig_df <- results %>%
      filter(p_val_2week < 0.05 | p_val_2month < 0.05 | p_val_6month < 0.05)
    sig_microbes <- sig_df$species
    #create new sig df
    taxa_relab_species_sig <- taxa_relab_species %>% 
      filter(rownames(.) %in% sig_microbes)

```

#Supplemental Figure 3: Heatmap of AC versus Taxa (family)
```{r data}

#define rmcorr function
rmcorr_test <- function(metadata,pid,df1,df2,p.adjust.method){
  if(all(sapply(list(rownames(df1),rownames(df2)),function(x) x==rownames(metadata)))){
    data<-cbind(metadata,df1,df2)
    rmcor<-NULL
    for (i in colnames(df1)){
      for (x in colnames(df2)){
        tryCatch({
        cor<-rmcorr(participant=get(pid),measure1=get(x),measure2=get(i),dataset=data)}, warning=function(w){})
        temp<-data.frame(i,x,cor$r,cor$p)
        rmcor<-rbind(rmcor,temp)}}
    rmcor["p.adj"]<-p.adjust(rmcor$cor.p, method=p.adjust.method) 
    names(rmcor)<-c("df1","df2","r","p","p.adj")
    return(rmcor)}else{
    warning("Rownames of inputs must match")}}

#prepare data frames
  Lipids_log2_1e6scaled_minimputated_sig_for_rmcorr <- lipids_log2_1e6scaled_minimputated_sig %>%
    t() %>%
    as.data.frame() %>%
    replace(is.na(.), 0) %>%
    rownames_to_column("sample_ID") %>%
    arrange(match(sample_ID, metadata$sample_ID)) %>%
    column_to_rownames("sample_ID")

  taxa_relab_family_sig_for_rmcorr <- taxa_relab_family_sig %>%
    t() %>%
    as.data.frame() %>%
    replace(is.na(.), 0) %>%
    rownames_to_column("sample_ID") %>%
    arrange(match(sample_ID, metadata$sample_ID)) %>%
    column_to_rownames("sample_ID")


#manually assign colors
family_colors <- c(
    "Enterobacteriaceae" = "#7a2e0a",
    "Lachnospiraceae" = "#06235E",
    "Oscillospiraceae" = "#e69f29",
    "Coriobacteriaceae" = "#cf3f4d",
    "Eubacteriales_unclassified" = "#139800")
lipid_group_colors <- c("AC" = "#004164")

#create empty list to store plots
plots <- list()

#loop through lipid groups
for (lipid_group in c("AC")) {
  #subset lipids for the current group
  lipid_subset <- as.data.frame(Lipids_log2_1e6scaled_minimputated_sig_for_rmcorr[, grep(paste0("^", lipid_group, "\\("), colnames(Lipids_log2_1e6scaled_minimputated_sig_for_rmcorr))])
  print(paste("Processing lipid group:", lipid_group)) 
  #skip groups with no or only one lipid
  if (is.null(lipid_subset) || ncol(lipid_subset) == 0 || all(is.na(lipid_subset)) || ncol(lipid_subset) <= 1) {
    print(paste("Skipping lipid group:", lipid_group, "because it has too few lipids"))
    next}
  #run rmcorr
  rmcor_Lipids_microbe <- rmcorr_test(
    metadata %>% column_to_rownames("sample_ID"), 
    "Recipient", 
    taxa_relab_family_sig_for_rmcorr, 
    lipid_subset, 
    "BH")
  print("Finished rmcor")
  #prepare data for heatmap
  microbe_Lipid_cor_plot <- rmcor_Lipids_microbe %>%
    dplyr::rename("microbe" = df1, "Lipid" = df2) %>%
    filter(!is.na(microbe)) %>%
    filter(Lipid %in% (filter(., p.adj <= 0.05))$Lipid) %>%
    select(-p, -p.adj) %>%
    group_by(Lipid) %>%
    complete(microbe, fill = list(Lipid = "drop", r = 0)) %>%
    spread(key = microbe, value = r, fill = 0) %>%
    column_to_rownames("Lipid") %>%
    select_if(~ sd(., na.rm = TRUE) > 0)  # remove zero-variation columns
  
  #check if plot is not empty
  if (nrow(microbe_Lipid_cor_plot) > 1 & ncol(microbe_Lipid_cor_plot) > 1) {
    #create annotation for columns (family of microbes)
    annotation_col <- data.frame(family = colnames(taxa_relab_family_sig_for_rmcorr))
    rownames(annotation_col) <- annotation_col$family
    # Create annotation for row (family of lipids)
    annotation_row <- lipid_group_key[lipid_group_key$Lipid %in% rownames(microbe_Lipid_cor_plot), ]
    annotation_row <- data.frame(Lipid_Group = annotation_row$Group, row.names = annotation_row$Lipid)
    # Combine both annotations into one list
    ann_colors <- list(
      Lipid_Group = lipid_group_colors,
      family = family_colors)

    print("finished preprocessing")
    
    #generate pheatmap for the current lipid group with both annotations
    pheatmap_plot <- pheatmap(microbe_Lipid_cor_plot,
                              number_color = "white",
                              cluster_rows = TRUE,
                              cluster_cols = TRUE,
                              clustering_distance_cols = "correlation",
                              color = c(rev(colorRampPalette(c("grey88", "dodgerblue", "dodgerblue3"), bias = 1)(30)),
                                        colorRampPalette(c("grey88", "red", "red3"), bias = 1)(30)),
                              annotation_row = annotation_row,
                              annotation_col = annotation_col,  
                              annotation_colors = ann_colors, 
                              main = paste("Correlation Heatmap of MEM sig Microbial Families versus Mem Sig Lipid (log2_1e6scaled_minimputated) Class:", lipid_group), 
                              legend = TRUE)
  
    #store the plot in the list using lipid group as the key
    plots[[lipid_group]] <- pheatmap_plot
    print(paste("finished plot for", lipid_group))} else {
    print(paste("No data to plot for lipid group:", lipid_group))
  }
}

# Save plots
output_dir <- "../Figures"

if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)}

for (lipid_group in names(plots)) {
  file_name <- paste0(output_dir, "/", "Fig_#_", lipid_group, "_Lipid_versus_Family_Heatmap.svg")  
  ggsave(
    filename = file_name,
    plot = plots[[lipid_group]],
    width = 8,
    height = 6,
    dpi = 300,
    limitsize = FALSE)}


```

#Fig 3A: Heatmaps of AC versus Taxa (Entero Species)
```{r data}

#prepare data frames
  Lipids_log2_1e6scaled_minimputated_sig_for_rmcorr <- lipids_log2_1e6scaled_minimputated %>%
    t() %>%
    as.data.frame() %>%
    replace(is.na(.), 0) %>%
    rownames_to_column("sample_ID") %>%
    arrange(match(sample_ID, metadata$sample_ID)) %>%
    column_to_rownames("sample_ID")

  taxa_for_rmcorr <- taxa_relab_enterobacteriaceae_species %>%
    t() %>%
    as.data.frame() %>%
    replace(is.na(.), 0) %>%
    rownames_to_column("sample_ID") %>%
    arrange(match(sample_ID, metadata$sample_ID)) %>%
    column_to_rownames("sample_ID")

# Manually assign colors
genus_colors <- c(
    "Citrobacter"   = "#052A2D",
    "Enterobacter"  = "#106749",
    "Escherichia"   = "#168246",
    "Klebsiella"    = "#1D9D3B",
    "Kluyvera"      = "#64BA56",
    "Phytobacter"   = "#8FC874",
    "Salmonella"    = "#D3E2B0")

#set lipid order
lipid_order_key <- read_excel("C:/Users/aganc/Desktop/Professional/Projects/FMT_Lipid_MCBA_Microbiome_Paper/Data/lipids/lipid_order_key.xlsx")

#create empty list to store plots
plots <- list()

#loop through lipid groups
for (lipid_group in c("AC")) {
  
  #subset lipids for the current group
  lipid_subset <- as.data.frame(Lipids_log2_1e6scaled_minimputated_sig_for_rmcorr[, grep(paste0("^", lipid_group, "\\("), colnames(Lipids_log2_1e6scaled_minimputated_sig_for_rmcorr))])
  print(paste("Processing lipid group:", lipid_group)) 
  
  #skip groups with no or only one lipid
  if (is.null(lipid_subset) || ncol(lipid_subset) == 0 || all(is.na(lipid_subset)) || ncol(lipid_subset) <= 1) {
    print(paste("Skipping lipid group:", lipid_group, "because it has too few lipids"))
    next}
  
  #run rmcorr
  rmcor_Lipids_microbe <- rmcorr_test(
    metadata  %>% column_to_rownames("sample_ID"), 
    "Recipient", 
    taxa_for_rmcorr, 
    lipid_subset, 
    "BH")
  
  print("Finished rmcor")
  
  #prepare data for heatmap
  microbe_Lipid_cor_plot <- rmcor_Lipids_microbe %>%
    dplyr::rename("microbe" = df1, "Lipid" = df2) %>%
    filter(!is.na(microbe)) %>%
    filter(Lipid %in% (filter(., p.adj <= 0.05))$Lipid) %>%
    select(-p, -p.adj) %>%
    group_by(Lipid) %>%
    complete(microbe, fill = list(Lipid = "drop", r = 0)) %>%
    spread(key = microbe, value = r, fill = 0) %>%
    column_to_rownames("Lipid") %>%
    select_if(~ sd(., na.rm = TRUE) > 0)  # remove zero-variation columns
  
  #check if MCBA_Lipid_cor_plot is not empty
  if (nrow(microbe_Lipid_cor_plot) > 1 & ncol(microbe_Lipid_cor_plot) > 1) {
    
    #define annotation
    annotation_col <- entero_genus_species_lookup %>%
      filter(Species %in% colnames(microbe_Lipid_cor_plot)) %>%
      column_to_rownames("Species") 
    #define ann_colors using genus_colors
    ann_colors <- list(Genus = genus_colors)
    
    #order lipids: 
      if (lipid_group == "AC") {
        #for AC, order by numeric 'length'
        lipid_order_present <- lipid_order_key %>%
          filter(str_starts(Lipid, lipid_group)) %>%
          filter(Lipid %in% rownames(microbe_Lipid_cor_plot)) %>%
          arrange(length) %>%
          pull(Lipid)
        #reorder heatmap rows
        microbe_Lipid_cor_plot <- microbe_Lipid_cor_plot[lipid_order_present, , drop = FALSE]
        
      } else if (lipid_group == "Cer") {
        #for Cer, order first by d_t (factor), then by Lipid name (optional)
        lipid_order_present <- lipid_order_key %>%
          filter(str_starts(Lipid, lipid_group)) %>%
          filter(Lipid %in% rownames(microbe_Lipid_cor_plot)) %>%
          mutate(d_t = factor(d_t, levels = c("d", "t"))) %>%
          arrange(d_t, Lipid) %>%
          pull(Lipid)
        
        #reorder heatmap rows
        microbe_Lipid_cor_plot <- microbe_Lipid_cor_plot[lipid_order_present, , drop = FALSE]
      }
    
    print("finished preprocessing")

    
    #generate pheatmap for the current lipid group with both annotations
    pheatmap_plot <- pheatmap(microbe_Lipid_cor_plot,
                              number_color = "white",
                              cluster_rows = FALSE,
                              cluster_cols = TRUE,
                              clustering_distance_cols = "correlation",
                              color = c(rev(colorRampPalette(c("grey88", "dodgerblue", "dodgerblue3"), bias = 1)(30)), 
                              colorRampPalette(c("grey88", "red", "red3"), bias = 1)(30)),
                              annotation_col = annotation_col,  
                              annotation_colors = ann_colors,  
                              main = paste("Correlation Heatmap of MEM sig Microbial Families versus Mem Sig Lipid (log2_1e6scaled_minimputated) Class:", lipid_group), 
                              legend = TRUE)
    
    #store the plot in the list using lipid group as the key
    plots[[lipid_group]] <- pheatmap_plot
    print(paste("finished plot for", lipid_group))} else {
    print(paste("No data to plot for lipid group:", lipid_group))
  }
}


# Save plots
output_dir <- "../Figures"

if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

for (lipid_group in names(plots)) {
  file_name <- paste0(output_dir, "/", "Fig_#_", lipid_group, "_Lipid_versus_Entero_Species_Heatmap_recolored.svg")  
  ggsave(
    filename = file_name,
    plot = plots[[lipid_group]],
    width = 8,
    height = 6,
    dpi = 300,
    limitsize = FALSE
  )
}


```
#Fig 3B: Heatmaps of AC and Cer versus Taxa (Lachno Species)
```{r data}

# Prepare data frames
  Lipids_log2_1e6scaled_minimputated_sig_for_rmcorr <- lipids_log2_1e6scaled_minimputated %>%
    t() %>%
    as.data.frame() %>%
    replace(is.na(.), 0) %>%
    rownames_to_column("sample_ID") %>%
    arrange(match(sample_ID, metadata$sample_ID)) %>%
    column_to_rownames("sample_ID")

  taxa_for_rmcorr <- taxa_relab_lachnospiraceae_species %>%
    t() %>%
    as.data.frame() %>%
    replace(is.na(.), 0) %>%
    rownames_to_column("sample_ID") %>%
    arrange(match(sample_ID, metadata$sample_ID)) %>%
    column_to_rownames("sample_ID")


# Manually assign colors
unique(lachno_genus_species_lookup$Genus)
genus_colors <- c(
    "Anaerobutyricum" = "#8c564b",
    "Anaerosporobacter" = "#ff7f0e",
    "Anaerostipes" = "#d62728",
    "Anaerotignum" = "#5254a3",
    "Blautia" = "#1f77b4",
    "Brotolimicola" = "#dbdb8d",
    "Cuneatibacter" = "#c49c94",
    "Coprococcus" = "#ffbb78",
    "Diplocloster" = "#8ca252",
    "Dorea" = "#9467bd",
    "Enterocloster" = "#c5b0d5",
    "Extibacter" = "#ff9896",
    "Eisenbergiella" = "#7f7f7f",
    "Faecalicatena" = "#637939",
    "Faecalimonas" = "#9c9ede",
    "Frisingicoccus" = "#7b4173",
    "Fusicatenibacter" = "#bcbd22",
    "Lachnoclostridium" = "#9edae5",
    "Lachnospira" = "#8c6d31",
    "Lachnospiraceae_unclassified" = "#17becf",
    "Mediterraneibacter" = "#2ca02c",
    "Merdimonas" = "#843c39",
    "Oliverpabstia" = "#e377c2",
    "Roseburia" = "#aec7e8",
    "Sellimonas" = "#98df8a",
    "Simiaoa" = "#b5cf6b",
    "Tyzzerella" = "#637939",
    "Wansuia" = "#393b79",
    "Zhenhengia" = "#f7b6d2",
    "Other" = "grey")

# Set lipid order
lipid_order_key <- read_excel("C:/Users/aganc/Desktop/Professional/Projects/FMT_Lipid_MCBA_Microbiome_Paper/Data/lipids/lipid_order_key.xlsx")

# Create empty list to store plots
plots <- list()

# Loop through lipid groups
for (lipid_group in c("AC")) {
  
  # Subset lipids for the current group
  lipid_subset <- as.data.frame(Lipids_log2_1e6scaled_minimputated_sig_for_rmcorr[, grep(paste0("^", lipid_group, "\\("), colnames(Lipids_log2_1e6scaled_minimputated_sig_for_rmcorr))])
  print(paste("Processing lipid group:", lipid_group)) 
  
  # Skip groups with no or only one lipid
  if (is.null(lipid_subset) || ncol(lipid_subset) == 0 || all(is.na(lipid_subset)) || ncol(lipid_subset) <= 1) {
    print(paste("Skipping lipid group:", lipid_group, "because it has too few lipids"))
    next}
  
  # Run rmcorr
  rmcor_Lipids_microbe <- rmcorr_test(
    metadata  %>% column_to_rownames("sample_ID"), 
    "Recipient", 
    taxa_for_rmcorr, 
    lipid_subset, 
    "BH")
  
  print("Finished rmcor")
  
  # Prepare data for heatmap
  microbe_Lipid_cor_plot <- rmcor_Lipids_microbe %>%
    dplyr::rename("microbe" = df1, "Lipid" = df2) %>%
    filter(!is.na(microbe)) %>%
    filter(Lipid %in% (filter(., p.adj <= 0.05))$Lipid) %>%
    select(-p, -p.adj) %>%
    group_by(Lipid) %>%
    complete(microbe, fill = list(Lipid = "drop", r = 0)) %>%
    spread(key = microbe, value = r, fill = 0) %>%
    column_to_rownames("Lipid") %>%
    select_if(~ sd(., na.rm = TRUE) > 0)  # remove zero-variation columns
  
  # Check if MCBA_Lipid_cor_plot is not empty
  if (nrow(microbe_Lipid_cor_plot) > 1 & ncol(microbe_Lipid_cor_plot) > 1) {
    
    # Define annotation
    annotation_col <- lachno_genus_species_lookup %>%
      filter(Species %in% colnames(microbe_Lipid_cor_plot)) %>%
      column_to_rownames("Species") 
    # Define ann_colors using genus_colors
    ann_colors <- list(Genus = genus_colors)
    
    #order lipids: 
      # --- Lipid ordering based on lipid_group ---
      if (lipid_group == "AC") {
        # For AC, order by numeric 'length'
        lipid_order_present <- lipid_order_key %>%
          filter(str_starts(Lipid, lipid_group)) %>%
          filter(Lipid %in% rownames(microbe_Lipid_cor_plot)) %>%
          arrange(length) %>%
          pull(Lipid)
        
        # Reorder heatmap rows
        microbe_Lipid_cor_plot <- microbe_Lipid_cor_plot[lipid_order_present, , drop = FALSE]
        
      } else if (lipid_group == "Cer") {
        # For Cer, order first by d_t (factor), then by Lipid name (optional)
        lipid_order_present <- lipid_order_key %>%
          filter(str_starts(Lipid, lipid_group)) %>%
          filter(Lipid %in% rownames(microbe_Lipid_cor_plot)) %>%
          mutate(d_t = factor(d_t, levels = c("d", "t"))) %>%
          arrange(d_t, Lipid) %>%
          pull(Lipid)
        
        # Reorder heatmap rows
        microbe_Lipid_cor_plot <- microbe_Lipid_cor_plot[lipid_order_present, , drop = FALSE]
      }
    
    print("finished preprocessing")

    
    # Generate pheatmap for the current lipid group with both annotations
    pheatmap_plot <- pheatmap(microbe_Lipid_cor_plot,
                              number_color = "white",
                              cluster_rows = FALSE,
                              cluster_cols = TRUE,
                              clustering_distance_cols = "correlation",
                              color = c(rev(colorRampPalette(c("grey88", "dodgerblue", "dodgerblue3"), bias = 1)(30)), 
                              colorRampPalette(c("grey88", "red", "red3"), bias = 1)(30)),
                              annotation_col = annotation_col,  # Use combined annotation
                              annotation_colors = ann_colors,  # Define colors for annotations
                              main = paste("Correlation Heatmap of MEM sig Microbial Families versus Mem Sig Lipid (log2_1e6scaled_minimputated) Class:", lipid_group), 
                              legend = TRUE)
    
    # Store the plot in the list using lipid group as the key
    plots[[lipid_group]] <- pheatmap_plot
    
    print(paste("finished plot for", lipid_group))} else {
    print(paste("No data to plot for lipid group:", lipid_group))
  }
}


# Save plots
output_dir <- "../Figures"

if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

for (lipid_group in names(plots)) {
  file_name <- paste0(output_dir, "/", "Fig_#_", lipid_group, "_Lipid_versus_Lachno_Species_Heatmap.svg")  
  ggsave(
    filename = file_name,
    plot = plots[[lipid_group]],
    width = 8,
    height = 6,
    dpi = 300,
    limitsize = FALSE
  )
}


```
#Statistical Test: Mixed Effects Modeling to determine pathways that are significant against time 
```{r data}

#melt and filter data
  #filter the metadata 
  filtered_metadata <- metadata %>%
    filter(time %in% c("Pre", "2 week", "2 month", "6 month"))
  #keep only the columns in the metabolite matrix that match filtered metadata
  data_long <- pathway_clean_unstrat_filtered %>%
    pivot_longer(cols = -pathway, names_to = "sample_ID", values_to = "value") %>%
    left_join(filtered_metadata, by = "sample_ID")

#get the unique list 
unique_pathways<- unique(data_long$pathway)

#initialize a dataframe
results <- data.frame(matrix(ncol = 4, nrow =0))
colnames(results) <- c('pathway', 'p_val_2week', 'p_val_2month', 'p_val_6month')

#loop through 
for(i in 1:length(unique_pathways)){
  #subset the data 
  data_sub <- subset(data_long, pathway == unique_pathways[i])
  model <- lmerTest::lmer(value ~ time + (1|Recipient), data = data_sub)
  summary <- summary(model)
  p_val_2week <- summary$coefficients[['time2 week', 'Pr(>|t|)']]
  p_val_2month <- summary$coefficients[['time2 month', 'Pr(>|t|)']]
  p_val_6month <- summary$coefficients[['time6 month', 'Pr(>|t|)']]
  #append results
  results[i, 'pathway'] <- unique_pathways[i]
  results[i, 'p_val_2week'] <- p_val_2week
  results[i, 'p_val_2month'] <- p_val_2month
  results[i, 'p_val_6month'] <- p_val_6month
}

#correct the p-values
results$p_val_2week <- p.adjust(results$p_val_2week, method = 'BH')
results$p_val_2month <- p.adjust(results$p_val_2month, method = 'BH')
results$p_val_6month <- p.adjust(results$p_val_6month, method = 'BH')

#subset significant 
sig_df <- results %>%
  filter(p_val_2week < 0.05 | p_val_2month < 0.05 | p_val_6month < 0.05)
sig_pathways <- sig_df$pathway

# Create new sig df
pathway_clean_unstrat_filtered_sig <- pathway_clean_unstrat_filtered %>%
  filter(pathway %in% sig_pathways)


```


#Fig 3A: Lipids versus amino acid pathway analyses
```{r data}

#filter signficant pathways to just amino acid related pathways
  amino_acid_pathways <- pathway_clean_unstrat_filtered_sig %>%
    rename(full_pathway = pathway) %>% 
    mutate(pathway_short_name = sub(":.*", "", full_pathway)) %>%  
    semi_join(Pathways_Category_Key %>%
        filter(broad_ontology %in% c(
          "Amino Acid Biosynthesis",
          "Amino Acid Degradation")), by = "pathway_short_name") %>%
    select(-pathway_short_name) %>%                 
    rename(pathway = full_pathway)
  
#run rmcorr on these against AC 
      #import keys
      pathway_full_to_short_name_key <- read.csv("C:/Users/aganc/Desktop/Professional/Projects/FMT_Lipid_MCBA_Microbiome_Paper/Tables/pathway_full_to_short_name_key.csv")
      Pathways_Category_Key <- read.xlsx("C:/Users/aganc/Desktop/Professional/Projects/FMT_Lipid_MCBA_Microbiome_Paper/Data/humann/Pathways_Category_Key.xlsx", sheetIndex = 1)
      
      #Prepare data frames
        Lipids_log2_1e6scaled_minimputated_for_rmcorr <- lipids_log2_1e6scaled_minimputated %>%
            t() %>%
            as.data.frame() %>%
            replace(is.na(.), 0) %>%
            rownames_to_column("sample_ID") %>%
            arrange(match(sample_ID, metadata$sample_ID)) %>%
            column_to_rownames("sample_ID")
          
        pathway_for_rmcorr <- amino_acid_pathways %>% column_to_rownames("pathway") %>% 
            t() %>%
            as.data.frame() %>% 
            mutate(across(everything(), as.numeric)) %>%  # convert all columns to numeric
            replace(is.na(.), 0) %>%
            mutate(across(everything(), ~log2(. + 1))) %>%  # log2 transform all numeric values
            rownames_to_column("sample_ID") %>%
            arrange(match(sample_ID, metadata$sample_ID)) %>%
            column_to_rownames("sample_ID") 
            
          #manually assign colors
          pathway_colors <- c(
            "Amino Acid Biosynthesis" = "#84d3ff",
            "Amino Acid Degradation" = "#004164")
          
          #set lipid order
          lipid_order_key <- read_excel("C:/Users/aganc/Desktop/Professional/Projects/FMT_Lipid_MCBA_Microbiome_Paper/Data/lipids/lipid_order_key.xlsx")
          
        #create empty list to store plots
          plots <- list()
          
        #loop through lipid groups
          for (lipid_group in c("AC")) {
            #subset lipids for the current group
            lipid_subset <- as.data.frame(Lipids_log2_1e6scaled_minimputated_sig_for_rmcorr[, grep(paste0("^", lipid_group, "\\("), colnames(Lipids_log2_1e6scaled_minimputated_sig_for_rmcorr))])
            print(paste("Processing lipid group:", lipid_group)) 
            
            #skip groups with no or only one lipid
            if (is.null(lipid_subset) || ncol(lipid_subset) == 0 || all(is.na(lipid_subset)) || ncol(lipid_subset) <= 1) {
              print(paste("Skipping lipid group:", lipid_group, "because it has too few lipids"))
              next}
            
            #run rmcorr
            rmcor_Lipids_pathway <- rmcorr_test(
              metadata %>% column_to_rownames("sample_ID"), 
              "Recipient", 
              pathway_for_rmcorr, 
              lipid_subset, 
              "BH")
            
            print("Finished rmcor")
            
            #prepare data for heatmap
            pathway_Lipid_cor_plot <- rmcor_Lipids_pathway %>%
              dplyr::rename("pathway" = df1, "Lipid" = df2) %>%
              filter(!is.na(pathway)) %>%
              filter(Lipid %in% (filter(., p.adj <= 0.05))$Lipid) %>% 
              select(-p, -p.adj) %>%
              group_by(Lipid) %>%
              complete(pathway, fill = list(Lipid = "drop", r = 0)) %>%
              spread(key = pathway, value = r, fill = 0) %>%
              column_to_rownames("Lipid") %>%
              select_if(~ sd(., na.rm = TRUE) > 0)  
            
            #check if MCBA_Lipid_cor_plot is not empty
            if (nrow(pathway_Lipid_cor_plot) > 1 & ncol(pathway_Lipid_cor_plot) > 1) {
              
              #define annotation
              annotation_col <- Pathways_Category_Key[, -c(2,3)] %>%
                filter(pathway %in% colnames(pathway_Lipid_cor_plot)) %>%
                column_to_rownames("pathway")  
              #define ann_colors 
              colnames(annotation_col) <- "pathway"
              ann_colors <- list(pathway = pathway_colors)
              
              #order lipids: 
              if (lipid_group == "AC") {
                lipid_order_present <- lipid_order_key %>%
                  filter(str_starts(Lipid, lipid_group)) %>%
                  filter(Lipid %in% rownames(pathway_Lipid_cor_plot)) %>%
                  arrange(length) %>%
                  pull(Lipid)
                
                #reorder heatmap rows
                pathway_Lipid_cor_plot <- pathway_Lipid_cor_plot[lipid_order_present, , drop = FALSE]} else if (lipid_group == "Cer") {
                #for Cer, order first by d_t (factor), then by Lipid name (optional)
                lipid_order_present <- lipid_order_key %>%
                  filter(str_starts(Lipid, lipid_group)) %>%
                  filter(Lipid %in% rownames(pathway_Lipid_cor_plot)) %>%
                  mutate(d_t = factor(d_t, levels = c("d", "t"))) %>%
                  arrange(d_t, Lipid) %>%
                  pull(Lipid)
                
                #reorder heatmap rows
                pathway_Lipid_cor_plot <- pathway_Lipid_cor_plot[lipid_order_present, , drop = FALSE]}
              
              #reorder heatmap columns by annotation
              ordered_cols <- annotation_col %>%
                rownames_to_column(var = "pathway_id") %>%
                mutate(broad_ontology = factor(pathway, levels = names(pathway_categories))) %>%
                arrange(broad_ontology) %>%
                pull(pathway_id)
              
              #apply column order
              pathway_Lipid_cor_plot <- pathway_Lipid_cor_plot[, ordered_cols, drop = FALSE]
              
              print("finished preprocessing")
              
              #generate pheatmap for the current lipid group with both annotations
              pheatmap_plot <- pheatmap(pathway_Lipid_cor_plot,
                                        number_color = "white",
                                        cluster_rows = FALSE,
                                        cluster_cols = TRUE,
                                        clustering_distance_cols = "correlation",
                                        color = c(rev(colorRampPalette(c("grey88", "dodgerblue", "dodgerblue3"), bias = 1)(30)), 
                                                  colorRampPalette(c("grey88", "red", "red3"), bias = 1)(30)),
                                        annotation_col = annotation_col,  # Use combined annotation
                                        annotation_colors = ann_colors,  # Define colors for annotations
                                        main = paste("Correlation Heatmap of MEM sig Microbial Families versus Mem Sig Lipid (log2_1e6scaled_minimputated) Class:", lipid_group), 
                                        fontsize_col = 3,
                                        legend = TRUE)
              
              #store the plot in the list using lipid group as the key
              plots[[lipid_group]] <- pheatmap_plot
              
              print(paste("finished plot for", lipid_group))} else {
                print(paste("No data to plot for lipid group:", lipid_group))
              }
          }
          
          #save plots
          output_dir <- "../Figures"
          
          if (!dir.exists(output_dir)) {
            dir.create(output_dir, recursive = TRUE)}
        
          for (lipid_group in names(plots)) {
            file_name <- paste0(output_dir, "/", "Fig_#_", lipid_group, "_Lipid_versus_amino_acid_pathways.svg")  
            ggsave(
              filename = file_name,
              plot = plots[[lipid_group]],
              width = 24,
              height = 18,
              dpi = 300,
              limitsize = FALSE)}
    
          
```         
      
#Fig 4B: Amino Acid biosynthesis and degradation plots
```{r data}

#biosynth pathways 
  #import df
  amino_acid_biosynth_pathways <- pathway_clean_unstrat_filtered_sig %>%
      rename(full_pathway = pathway) %>% 
      mutate(pathway_short_name = sub(":.*", "", full_pathway)) %>%  
      semi_join(Pathways_Category_Key %>%
          filter(broad_ontology %in% c(
            "Amino Acid Biosynthesis")), by = "pathway_short_name") %>%
      select(-pathway_short_name) %>%                 
      rename(pathway = full_pathway)
  
  #prepare the stratified dataset
  pathway_sig_strat <- pathway_clean %>%
    filter(grepl("\\|", pathway)) %>%
    filter(sub("\\|.*", "", pathway) %in% unique(amino_acid_biosynth_pathways$pathway))
  
  #separate pathway and taxonomy
  pathway_long <- pathway_sig_strat %>%
    separate(pathway, into = c("pathway", "taxonomy"), sep = "\\|") %>%
    mutate(
      full_taxonomy = taxonomy,
      genus = str_extract(taxonomy, "g__[^\\.]+")) %>%
    relocate(full_taxonomy, .after = taxonomy) %>%
    relocate(genus, .after = full_taxonomy) %>%
    mutate(genus = if_else(is.na(genus), "unclassified", genus))
  
  #sum abundances across samples
  genus_sum <- pathway_long %>%
    group_by(genus) %>%
    summarise(across(where(is.numeric), ~sum(.x, na.rm = TRUE)), .groups = "drop")
  
  #determine top 20 genera 
  top20_genera <- genus_sum %>%
    pivot_longer(-genus, names_to = "sample_ID", values_to = "abundance") %>%
    group_by(genus) %>%
    summarise(total_abundance = sum(abundance, na.rm = TRUE), .groups = "drop") %>%
    arrange(desc(total_abundance)) %>%
    slice_head(n = 20) %>%
    pull(genus)
  
  #prepare plotting dataset
  genus_abs_with_time_lumped <- genus_sum %>%
    pivot_longer(-genus, names_to = "sample_ID", values_to = "abundance") %>%
    mutate(genus = ifelse(genus %in% top20_genera, genus, "Other")) %>%
    group_by(genus, sample_ID) %>%
    summarise(abundance = sum(abundance, na.rm = TRUE), .groups = "drop") %>%
    left_join(metadata %>% select(sample_ID, time), by = "sample_ID") %>%
    mutate(time = factor(time, levels = c("Pre", "2 week", "2 month", "6 month"))) %>%
    arrange(time, sample_ID)
  
  top20_genera
  
  #order taxa
  desired_order <- c(
  "g__Bifidobacterium", 
  "g__Collinsella", 
  "g__Bacteroides",
  "g__Acidaminococcus",  
  "g__Anaerostipes", 
  "g__Blautia", 
  "g__Dorea",  
  "g__Enterococcus",  
  "g__Erysipelatoclostridium", 
  "g__Eubacterium", 
  "g__Faecalibacterium",    
  "g__Fusicatenibacter", 
  "g__Lachnospiraceae_unclassified",
  "g__Ruminococcus",                
  "g__Streptococcus", 
  "g__Citrobacter",
  "g__Escherichia", 
  "g__Klebsiella",   
  "g__Akkermansia",  
  "unclassified", 
  "Other")
  
  #apply order
  genus_abs_with_time_lumped <- genus_abs_with_time_lumped %>%
    mutate(genus = factor(genus, levels = desired_order))
  
  #define pallete
  palette_ordered <- c("#004164","#0081c6","#22b1ff","#84d3ff","#374f38","#719673","#a5c0a6","#d3e0d5","#e64051"
                       ,"#f08e98","#f5bcc1","#a46909","#df8f0b","#f7bc59","#874691"
                       ,"#a964b3","#bc87c4","#dbbddf","#b03900","#ff5402","#ff8f59","#ffb591","#1c0488"
                       ,"#2e06e6","#8a73fb","#d8cffe")
  
  #summarize abundance across genera per sample, per time
  sample_totals <- genus_abs_with_time_lumped %>%
    group_by(sample_ID, time) %>%
    summarise(total_abundance = sum(abundance, na.rm = TRUE), .groups = "drop")
  max_abundance <- max(sample_totals$total_abundance)
  max_abundance
  
  #plot
  biosynth_plot <- ggplot(genus_abs_with_time_lumped, aes(x = sample_ID, y = abundance, fill = genus)) +
    geom_bar(stat = "identity") +
    facet_wrap(~ time, scales = "free_x", nrow = 1) +
    scale_fill_manual(values = palette_ordered) +
    theme_minimal() +
    theme(
      strip.background = element_blank(),
      strip.text = element_text(face = "bold", size = 5),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      plot.title = element_text(hjust = 0.5, size = 15, face = "bold"), 
      legend.position = "right") +
    labs(
      x = "Sample ID",
      y = "Abundance (CPM)",
      fill = "Genus",
      title = "Amino Acid Biosynthesis Pathway Contributors")
   
  
  biosynth_plot
  
  #export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_AminoAcid_Biosynthesis_Pathway_Contributors.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = biosynth_plot,  # Use the combined plot
    width = 8,     # Width of the plot in inches
    height = 6,    # Height of the plot in inches
    dpi = 300,     # Resolution (dots per inch)
    limitsize = FALSE)  # Prevent clipping of larger legends or labels 





#degrad pathways
  #import df
  amino_acid_biosynth_pathways <- pathway_clean_unstrat_filtered_sig %>%
      rename(full_pathway = pathway) %>% 
      mutate(pathway_short_name = sub(":.*", "", full_pathway)) %>% 
      semi_join(Pathways_Category_Key %>%
          filter(broad_ontology %in% c(
            "Amino Acid Degradation")), by = "pathway_short_name") %>%
      select(-pathway_short_name) %>%                 
      rename(pathway = full_pathway)
  
  #pPrepare the df
  pathway_sig_strat <- pathway_clean %>%
    filter(grepl("\\|", pathway)) %>%
    filter(sub("\\|.*", "", pathway) %in% unique(amino_acid_biosynth_pathways$pathway))
  
  #separate pathway and taxonomy
  pathway_long <- pathway_sig_strat %>%
    separate(pathway, into = c("pathway", "taxonomy"), sep = "\\|") %>%
    mutate(
      full_taxonomy = taxonomy,
      genus = str_extract(taxonomy, "g__[^\\.]+")) %>%
    relocate(full_taxonomy, .after = taxonomy) %>%
    relocate(genus, .after = full_taxonomy) %>%
    mutate(genus = if_else(is.na(genus), "unclassified", genus))
  
  #sum abundances across samples
  genus_sum <- pathway_long %>%
    group_by(genus) %>%
    summarise(across(where(is.numeric), ~sum(.x, na.rm = TRUE)), .groups = "drop")
  
  #determine top 20 genera
  top20_genera <- genus_sum %>%
    pivot_longer(-genus, names_to = "sample_ID", values_to = "abundance") %>%
    group_by(genus) %>%
    summarise(total_abundance = sum(abundance, na.rm = TRUE), .groups = "drop") %>%
    arrange(desc(total_abundance)) %>%
    slice_head(n = 20) %>%
    pull(genus)
  
  #prepare plotting dataset
  genus_abs_with_time_lumped <- genus_sum %>%
    pivot_longer(-genus, names_to = "sample_ID", values_to = "abundance") %>%
    mutate(genus = ifelse(genus %in% top20_genera, genus, "Other")) %>%
    group_by(genus, sample_ID) %>%
    summarise(abundance = sum(abundance, na.rm = TRUE), .groups = "drop") %>%
    left_join(metadata %>% select(sample_ID, time), by = "sample_ID") %>%
    mutate(time = factor(time, levels = c("Pre", "2 week", "2 month", "6 month"))) %>%
    arrange(time, sample_ID)
  
  top20_genera
  
  #define order
  desired_order <- c(
    "g__Bacteroides",
    "g__Parabacteroides",
    "g__Alistipes",
    "g__Dysgonomonas",
    "g__Fusobacterium",
    "g__Acidaminococcus",
    "g__Blautia",
    "g__Enterococcus",
    "g__Streptococcus",
    "g__Firmicutes_unclassified",
    "g__Citrobacter",
    "g__Enterobacter",
    "g__Escherichia",
    "g__Klebsiella",
    "g__Kluyvera",
    "g__Salmonella",
    "unclassified",
    "Other")
  
  #apply order
  genus_abs_with_time_lumped <- genus_abs_with_time_lumped %>%
    mutate(genus = factor(genus, levels = desired_order))
  
  #define pallete
  palette_ordered <- c("#004164","#0081c6","#22b1ff","#84d3ff","#374f38","#719673","#a5c0a6","#d3e0d5","#e64051"
                       ,"#f08e98","#f5bcc1","#a46909","#df8f0b","#f7bc59","#874691"
                       ,"#a964b3","#bc87c4","#dbbddf","#b03900","#ff5402","#ff8f59","#ffb591","#1c0488"
                       ,"#2e06e6","#8a73fb","#d8cffe")
  
  #summarize abundance across genera per sample, per time
  sample_totals <- genus_abs_with_time_lumped %>%
    group_by(sample_ID, time) %>%
    summarise(total_abundance = sum(abundance, na.rm = TRUE), .groups = "drop")
  max_abundance <- max(sample_totals$total_abundance)
  max_abundance
  
  #plot
  degrad_plot <- ggplot(genus_abs_with_time_lumped, aes(x = sample_ID, y = abundance, fill = genus)) +
    geom_bar(stat = "identity") +
    facet_wrap(~ time, scales = "free_x", nrow = 1) +
    scale_fill_manual(values = palette_ordered) +
    theme_minimal() +
    theme(
      strip.background = element_blank(),
      strip.text = element_text(face = "bold", size = 5),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      plot.title = element_text(hjust = 0.5, size = 15, face = "bold"), 
      legend.position = "right") +
    labs(
      x = "Sample ID",
      y = "Abundance (CPM)",
      fill = "Genus",
      title = "Amino Acid Degrdation Pathway Contributors")
   
  
  degrad_plot

  #export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_AminoAcid_Degradation_Pathway_Contributors.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = degrad_plot,  # Use the combined plot
    width = 8,     # Width of the plot in inches
    height = 6,    # Height of the plot in inches
    dpi = 300,     # Resolution (dots per inch)
    limitsize = FALSE)  # Prevent clipping of larger legends or labels 


```      
      
      
      
      
      
      
      
      
      
      
      
      
          
#Fig Supplemental 4: Lipid biosynthesis and degradation taxa contirbutions
```{r data}

#biosynth 
  #subset df
  amino_acid_biosynth_pathways <- pathway_clean_unstrat_filtered_sig %>%
      rename(full_pathway = pathway) %>% 
      mutate(pathway_short_name = sub(":.*", "", full_pathway)) %>%  
      semi_join(Pathways_Category_Key %>%
          filter(broad_ontology %in% c(
            "Lipid Biosynthesis")), by = "pathway_short_name") %>%
      select(-pathway_short_name) %>%                 
      rename(pathway = full_pathway) 
  
  #prepare the stratified dataset
  pathway_sig_strat <- pathway_clean %>%
    filter(grepl("\\|", pathway)) %>%
    filter(sub("\\|.*", "", pathway) %in% unique(amino_acid_biosynth_pathways$pathway))
  
  #separate pathway and taxonomy
  pathway_long <- pathway_sig_strat %>%
    separate(pathway, into = c("pathway", "taxonomy"), sep = "\\|") %>%
    mutate(
      full_taxonomy = taxonomy,
      genus = str_extract(taxonomy, "g__[^\\.]+")) %>%
    relocate(full_taxonomy, .after = taxonomy) %>%
    relocate(genus, .after = full_taxonomy) %>%
    mutate(genus = if_else(is.na(genus), "unclassified", genus))
  
  #sum abundances across samples 
  genus_sum <- pathway_long %>%
    group_by(genus) %>%
    summarise(across(where(is.numeric), ~sum(.x, na.rm = TRUE)), .groups = "drop")
  
  #determine top 20 genera 
  top20_genera <- genus_sum %>%
    pivot_longer(-genus, names_to = "sample_ID", values_to = "abundance") %>%
    group_by(genus) %>%
    summarise(total_abundance = sum(abundance, na.rm = TRUE), .groups = "drop") %>%
    arrange(desc(total_abundance)) %>%
    slice_head(n = 20) %>%
    pull(genus)
  
  #prepare plotting dataset
  genus_abs_with_time_lumped <- genus_sum %>%
    pivot_longer(-genus, names_to = "sample_ID", values_to = "abundance") %>%
    mutate(genus = ifelse(genus %in% top20_genera, genus, "Other")) %>%
    group_by(genus, sample_ID) %>%
    summarise(abundance = sum(abundance, na.rm = TRUE), .groups = "drop") %>%
    left_join(metadata %>% select(sample_ID, time), by = "sample_ID") %>%
    mutate(time = factor(time, levels = c("Pre", "2 week", "2 month", "6 month"))) %>%
    arrange(time, sample_ID)
  
  top20_genera
  
  #specify manual order for genera
  desired_order <- c(
    #Actinobacteriota 
    "g__Bifidobacterium",
    #Bacteroidota
    "g__Bacteroides",
    "g__Parabacteroides",
    "g__Alistipes",
    #Firmicutes
    "g__Blautia",
    "g__Fusicatenibacter",
    "g__Ruminococcaceae_unclassified",
    "g__Lactobacillus",
    "g__Enterococcus",
    "g__Streptococcus",
    "g__Pediococcus",
    "g__Veillonella",
    #Proteobacteria
    "g__Citrobacter",
    "g__Enterobacter",
    "g__Escherichia",
    "g__Klebsiella",
    "g__Kluyvera",
    "g__Salmonella",
    "g__Sutterella",
    #Unclassified / Other
    "unclassified",
    "Other")
  
  #apply order to factor
  genus_abs_with_time_lumped <- genus_abs_with_time_lumped %>%
    mutate(genus = factor(genus, levels = desired_order))
  
  #define palette
  palette_ordered <- c("#004164","#0081c6","#22b1ff","#84d3ff","#374f38","#719673","#a5c0a6","#d3e0d5","#e64051"
                       ,"#f08e98","#f5bcc1","#a46909","#df8f0b","#f7bc59","#874691"
                       ,"#a964b3","#bc87c4","#dbbddf","#b03900","#ff5402","#ff8f59","#ffb591","#1c0488"
                       ,"#2e06e6","#8a73fb","#d8cffe")
  
  #summarize
  sample_totals <- genus_abs_with_time_lumped %>%
    group_by(sample_ID, time) %>%
    summarise(total_abundance = sum(abundance, na.rm = TRUE), .groups = "drop")
  max_abundance <- max(sample_totals$total_abundance)
  max_abundance
  
  #plot
  biosynth_plot <- ggplot(genus_abs_with_time_lumped, aes(x = sample_ID, y = abundance, fill = genus)) +
    geom_bar(stat = "identity") +
    facet_wrap(~ time, scales = "free_x", nrow = 1) +
    scale_fill_manual(values = palette_ordered) +
    theme_minimal() +
    theme(
      strip.background = element_blank(),
      strip.text = element_text(face = "bold", size = 5),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      plot.title = element_text(hjust = 0.5, size = 15, face = "bold"), 
      legend.position = "right") +
    labs(
      x = "Sample ID",
      y = "Abundance (CPM)",
      fill = "Genus",
      title = "Lipid Biosynthesis Pathway Contributors")
   
  
  biosynth_plot
  
  #export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_Lipid_Biosynthesis_Pathway_Contributors.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = biosynth_plot,  
    width = 8,   
    height = 6, 
    dpi = 300,     
    limitsize = FALSE)  


#degrad pathways
  #subset df 
  amino_acid_biosynth_pathways <- pathway_clean_unstrat_filtered_sig %>%
      rownames_to_column(var = "full_pathway") %>%
      mutate(pathway_short_name = sub(":.*", "", full_pathway)) %>%  
      semi_join(Pathways_Category_Key %>%
          filter(broad_ontology %in% c(
            "Lipid Degradation")), by = "pathway_short_name") %>%
      select(-pathway_short_name) %>%                 
      rename(pathway = full_pathway)
  
  #prepare the stratified dataset
  pathway_sig_strat <- pathway_clean %>%
    filter(grepl("\\|", pathway)) %>%
    filter(sub("\\|.*", "", pathway) %in% unique(rownames(pathway_clean_unstrat_filtered_sig))) %>%
    filter(sub("\\|.*", "", pathway) %in% unique(amino_acid_biosynth_pathways$pathway))
  
  #separate pathway and taxonomy
  pathway_long <- pathway_sig_strat %>%
    separate(pathway, into = c("pathway", "taxonomy"), sep = "\\|") %>%
    mutate(
      full_taxonomy = taxonomy,
      genus = str_extract(taxonomy, "g__[^\\.]+")) %>%
    relocate(full_taxonomy, .after = taxonomy) %>%
    relocate(genus, .after = full_taxonomy) %>%
    mutate(genus = if_else(is.na(genus), "unclassified", genus))
  
  #sum abundances across samples 
  genus_sum <- pathway_long %>%
    group_by(genus) %>%
    summarise(across(where(is.numeric), ~sum(.x, na.rm = TRUE)), .groups = "drop")
  
  #determine top 20 genera 
  top20_genera <- genus_sum %>%
    pivot_longer(-genus, names_to = "sample_ID", values_to = "abundance") %>%
    group_by(genus) %>%
    summarise(total_abundance = sum(abundance, na.rm = TRUE), .groups = "drop") %>%
    arrange(desc(total_abundance)) %>%
    slice_head(n = 20) %>%
    pull(genus)
  
  #prepare plotting dataset
  genus_abs_with_time_lumped <- genus_sum %>%
    pivot_longer(-genus, names_to = "sample_ID", values_to = "abundance") %>%
    mutate(genus = ifelse(genus %in% top20_genera, genus, "Other")) %>%
    group_by(genus, sample_ID) %>%
    summarise(abundance = sum(abundance, na.rm = TRUE), .groups = "drop") %>%
    left_join(metadata %>% select(sample_ID, time), by = "sample_ID") %>%
    mutate(time = factor(time, levels = c("Pre", "2 week", "2 month", "6 month"))) %>%
    arrange(time, sample_ID)
  
  top20_genera
  
  desired_order <- c(
    "g__Coprococcus",
    "g__Citrobacter",
    "g__Enterobacter",
    "g__Escherichia",
    "g__Klebsiella",
    "g__Kluyvera",
    "g__Morganella",
    "g__Proteus",
    "g__Salmonella",
    "unclassified",
    "Other")
  
  #apply order
  genus_abs_with_time_lumped <- genus_abs_with_time_lumped %>%
    mutate(genus = factor(genus, levels = desired_order))
  
  #define pallete
  palette_ordered <- c("#004164","#0081c6","#22b1ff","#84d3ff","#374f38","#719673","#a5c0a6","#d3e0d5","#e64051"
                       ,"#f08e98","#f5bcc1","#a46909","#df8f0b","#f7bc59","#874691"
                       ,"#a964b3","#bc87c4","#dbbddf","#b03900","#ff5402","#ff8f59","#ffb591","#1c0488"
                       ,"#2e06e6","#8a73fb","#d8cffe")
  
  #summarize
  sample_totals <- genus_abs_with_time_lumped %>%
    group_by(sample_ID, time) %>%
    summarise(total_abundance = sum(abundance, na.rm = TRUE), .groups = "drop")
  max_abundance <- max(sample_totals$total_abundance)
  max_abundance
  
  #plot
  degrad_plot <- ggplot(genus_abs_with_time_lumped, aes(x = sample_ID, y = abundance, fill = genus)) +
    geom_bar(stat = "identity") +
    facet_wrap(~ time, scales = "free_x", nrow = 1) +
    scale_fill_manual(values = palette_ordered) +
    theme_minimal() +
    theme(
      strip.background = element_blank(),
      strip.text = element_text(face = "bold", size = 5),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      plot.title = element_text(hjust = 0.5, size = 15, face = "bold"), 
      legend.position = "right") +
    labs(
      x = "Sample ID",
      y = "Abundance (CPM)",
      fill = "Genus",
      title = "Lipid Degradation Pathway Contributors")
   
  
  degrad_plot
  
  #export figure
  #select dir
  output_dir <- "../Figures"
  # Generate the file name for the plot
  file_name <- paste0(output_dir, "/", "Fig_#_Lipid_Degradation_Pathway_Contributors.svg")  
  # Save the plot
  ggsave(
    filename = file_name,
    plot = degrad_plot, 
    width = 8,     
    height = 6,    
    dpi = 300,    
    limitsize = FALSE)  


```
          
      

